nav2_collision_monitor/src/polygon_source.cpp

```diff
@@ -0,0 +1,172 @@
+  }
+
+  std::string source_topic;
+
+  getParameters(source_topic);
+
+  rclcpp::QoS qos = rclcpp::SensorDataQoS();  // set to default
+  data_sub_ = node->create_subscription<nav2_msgs::msg::PolygonsArray>(
+    source_topic, qos,
+    std::bind(&PolygonSource::dataCallback, this, std::placeholders::_1));
+}
+
+void PolygonSource::getData(
+  const rclcpp::Time & curr_time,
+  std::vector<Point> & data) const
+{
+  // Ignore data from the source if it is not being published yet or
+  // not published for a long time
+  if (data_ == nullptr || data_->polygons.empty()) {
+    return;
+  }
+  // get the oldest time stamp from the polygon array
+  rclcpp::Time oldest_stamp = rclcpp::Time(data_->polygons[0].header.stamp);
+  for (const auto & polygon : data_->polygons) {
+    if (rclcpp::Time(polygon.header.stamp) < oldest_stamp) {
+      oldest_stamp = rclcpp::Time(polygon.header.stamp);
+    }
+  }
+  if (!sourceValid(oldest_stamp, curr_time)) {
+    return;
+  }
+
+  tf2::Stamped<tf2::Transform> tf_transform;
+
+  for (const auto & polygon : data_->polygons) {
+    if (base_shift_correction_) {
+      // Obtaining the transform to get data from source frame and time where it was received
+      // to the base frame and current time
+      if (
+        !nav2_util::getTransform(
+          polygon.header.frame_id, polygon.header.stamp,
+          base_frame_id_, curr_time, global_frame_id_,
+          transform_tolerance_, tf_buffer_, tf_transform))
+      {
+        return;
+      }
+    } else {
+      // Obtaining the transform to get data from source frame to base frame without time shift
+      // considered. Less accurate but much more faster option not dependent on state estimation
+      // frames.
+      if (
+        !nav2_util::getTransform(
+          polygon.header.frame_id, base_frame_id_,
+          transform_tolerance_, tf_buffer_, tf_transform))
+      {
+        return;
+      }
+    }
+    geometry_msgs::msg::PolygonStamped poly_out;
+    geometry_msgs::msg::TransformStamped tf = tf2::toMsg(tf_transform);
+    tf2::doTransform(polygon, poly_out, tf);
+    convertPolygonStampedToPoints(poly_out, data);
+  }
+
+}
+void PolygonSource::convertPolygonStampedToPoints(
+  const geometry_msgs::msg::PolygonStamped & polygon, std::vector<Point> & data) const
+{
+  // Iterate over the vertices of the polygon
+  for (size_t i = 0; i < polygon.polygon.points.size(); ++i) {
+    const auto & current_point = polygon.polygon.points[i];
+    const auto & next_point = polygon.polygon.points[(i + 1) % polygon.polygon.points.size()];
+
+    // Calculate the distance between the current and next points
+    double segment_length =
+      std::hypot(next_point.x - current_point.x, next_point.y - current_point.y);
+
+    // Calculate the number of points to sample in the current segment
+    size_t num_points_in_segment =
+      std::max(static_cast<size_t>(segment_length / sampling_distance_), static_cast<size_t>(1));
```

Comment:
Is not just `1` w/o static_cast won't be suitable here?

Reply:
probably, GPT generated that part :p

Reply:
nope, fails:

```
/code/ros2_ws/src/navigation2/nav2_collision_monitor/src/polygon_source.cpp:125:15: error: no matching function for call to ‘max(size_t, int)’
  125 |       std::max(static_cast<size_t>(segment_length / sampling_distance_), 1);
```

Reply:
It doesn't need to be size_t.

Also, please add a comment anywhere generative AI was used to generate code. I'm working on a proposal around that and it would be good practice for me to see if my suggestions are helpful :-)